hedera-mirror-rest/stateproof.js

```diff
@@ -0,0 +1,265 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const utils = require('./utils');
+const config = require('./config');
+const constants = require('./constants');
+const EntityId = require('./entityId');
+const TransactionId = require('./transactionId');
+const s3client = require('./s3client');
+const { DbError } = require('./errors/dbError');
+const { NotFoundError } = require('./errors/notFoundError');
+const { FileDownloadError } = require('./errors/fileDownloadError');
+
+/**
+ * Get the consensus_ns of the transaction. Throws exception if no such successful transaction found or multiple such
+ * transactions found.
+ * @param {TransactionId} transactionId
+ * @returns {Promise<String>} consensus_ns of the successful transaction if found
+ */
+let getSuccessfulTransactionConsensusNs = async (transactionId) => {
+  const sqlParams = [transactionId.getEntityId().getEncodedId(), transactionId.getValidStartNs()];
+  const sqlQuery = `SELECT consensus_ns
+       FROM transaction
+       WHERE payer_account_id = ?
+         AND valid_start_ns = ?
+         AND result = 22`; // only the successful transaction
+  const pgSqlQuery = utils.convertMySqlStyleQueryToPostgres(sqlQuery, sqlParams);
+  if (logger.isTraceEnabled()) {
+    logger.trace(`getSuccessfulTransactionConsensusNs: ${pgSqlQuery}, ${JSON.stringify(sqlParams)}`);
+  }
+
+  let result;
+  try {
+    result = await pool.query(pgSqlQuery, sqlParams);
+  } catch (err) {
+    throw new DbError(err.message);
+  }
+
+  const { rows } = result;
+  if (_.isEmpty(rows)) {
+    throw new NotFoundError('Transaction not found');
+  } else if (rows.length > 1) {
+    throw new DbError('Invalid state, more than one transactions found');
+  }
+
+  return _.first(rows).consensus_ns;
+};
+
+/**
+ * Get the RCD file name where consensusNs is in the range [consensus_start, consensus_end]. Throws exception if no
+ * such RCD file found.
+ * @param {string} consensusNs consensus timestamp within the range of the RCD file to search
+ * @returns {Promise<String>} RCD file name
+ */
+let getRCDFileNameByConsensusNs = async (consensusNs) => {
+  const sqlParams = [consensusNs, consensusNs];
+  const sqlQuery = `SELECT name
+       FROM record_file
+       WHERE consensus_start <= ?
+         AND consensus_end >= ?`;
+  const pgSqlQuery = utils.convertMySqlStyleQueryToPostgres(sqlQuery, sqlParams);
+  if (logger.isTraceEnabled()) {
+    logger.trace(`getRCDFileNameByConsensusNs: ${pgSqlQuery}, ${JSON.stringify(sqlParams)}`);
+  }
+
+  let result;
+  try {
+    result = await pool.query(pgSqlQuery, sqlParams);
+  } catch (err) {
+    throw new DbError(err.message);
+  }
+
+  const { rows } = result;
+  if (_.isEmpty(rows)) {
+    throw new NotFoundError('No matching RCD file found');
+  } else if (rows.length > 1) {
+    throw new DbError('Invalid state, more than one RCD files found');
```

Comment:
nit: `files` can be singular
```suggestion
    throw new DbError('Invalid state, more than one RCD file found');
```

Reply:
updated.